include/aws/crt/http/HttpConnection.h

```diff
@@ -0,0 +1,356 @@
+                HttpConnection &operator=(HttpConnection &&) = delete;
+
+                /**
+                 * Make a new client initiated request on this connection.
+                 *
+                 * If you take a reference to the return value, the memory and resources for the connection
+                 * and stream will not be cleaned up until you release it. You can however, release the reference
+                 * as soon as you don't need it anymore. The internal reference count ensures the resources will
+                 * not be freed until the stream is completed.
+                 *
+                 * Returns an instance of HttpStream upon success and nullptr on failure.
+                 */
+                std::shared_ptr<HttpStream> NewStream(const HttpRequestOptions &requestOptions) noexcept;
+
+                /**
+                 * Initiate a shutdown of the connection. Sometimes, connections are persistent and you want
+                 * to close them before shutting down your application or whatever is consuming this interface.
+                 *
+                 * Assuming `OnConnectionShutdown` has not already been invoked, it will be invoked as a result of this
+                 * call. It is safe to release your reference to this object after calling this function.
+                 */
+                bool Close() noexcept;
+
+                int LastError() const noexcept { return m_lastError; }
+
+              private:
+                HttpConnection(aws_http_connection *m_connection, Allocator *allocator) noexcept;
+                aws_http_connection *m_connection;
+                Allocator *m_allocator;
+                int m_lastError;
+
+                friend class HttpClient;
+            };
+
+            /**
+             * Represents an Http Client. `onConnectionSetup` will be invoked upon an outgoing connection,
+             * and `onConnectionShutdown` will be invoked when a connection is shutdown.
+             */
+            class AWS_CRT_CPP_API HttpClient final
+            {
+              public:
+                /**
+                 * Initializes client with a bootstrap object. `initialWindowSize` will set the TCP read window
+                 * allowed for Http 1.1 connections and Initial Windows for H2 connections.
+                 */
+                HttpClient(
+                    Io::ClientBootstrap *bootstrap,
+                    std::size_t initialWindowSize = SIZE_MAX,
+                    Allocator *allocator = DefaultAllocator()) noexcept;
+
+                /**
+                 * Establish a new Https Connection to hostName:port, using `socketOptions` for tcp options and
+                 * `tlsConnOptions` for TLS/SSL options.
+                 *
+                 * returns true on success, and false on failure. If false is returned, `onConnectionSetup` will not
+                 * be invoked.
+                 */
+                bool NewConnection(
```

Comment:
Given that it doesn't return like the rest of the factory patterns, perhaps this would better be named StartConnection? BeginConnection? CreateConnection?